File: java/78. Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        populate(nums, result, 0, new ArrayList<>());
        return result;
    }
    private void populate(
        int[] nums,
        List<List<Integer>> result,
        int index,
        List<Integer> current
    ) {
        if (index >= nums.length) {
            result.add(new ArrayList<>(current));
            return;
        }
        populate(nums, result, index+1, current);
        current.add(nums[index]);
        populate(nums, result, index+1, current);
        current.remove(current.size()-1);
    }
}

File: python/78. Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        current = []
        def populate(index):
            nonlocal result, current, nums
            if index >= len(nums):
                result.append(current.copy())
                return
            populate(index+1)
            current.append(nums[index])
            populate(index+1)
            current.pop()
        populate(0)
        return result